### ISSUE TITLE
Canvas - 500 errors (appears to be related to RCE API)

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#1362]
A user has reported that the current version of TurnKey Canvas is experiencing issues, namely this message:

```
Page Error
Something broke unexpectedly.
If you have a moment, click here to tell us what happened.
```

Within the logs, there are 500 errors being noted. The stacktrace in the database (full error messages are stored in the DB, rather than the log) shows more info:

```
       message        |                                                                                         backtrace                                                                                         
----------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 key must be 32 bytes | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/json-jwt-1.9.4/lib/json/jwe.rb:33:in `key='                                                                                                +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/json-jwt-1.9.4/lib/json/jwe.rb:33:in `encrypt!'                                                                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/json-jwt-1.9.4/lib/json/jwt.rb:40:in `encrypt'                                                                                             +
                      | /var/www/canvas/lib/canvas/security.rb:147:in `create_encrypted_jwt'                                                                                                                     +
                      | /var/www/canvas/lib/canvas/security/services_jwt.rb:71:in `generate'                                                                                                                     +
                      | /var/www/canvas/lib/canvas/security/services_jwt.rb:95:in `for_user'                                                                                                                     +
                      | /var/www/canvas/lib/services/rich_content.rb:24:in `env_for'                                                                                                                             +
                      | /var/www/canvas/app/controllers/application_controller.rb:198:in `rce_js_env'                                                                                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:413:in `block in make_lambda'                                                        +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:178:in `block (2 levels) in halting_and_conditional'                                 +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/abstract_controller/callbacks.rb:12:in `block (2 levels) in <module:Callbacks>'                                     +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:179:in `block in halting_and_conditional'                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:507:in `block in invoke_before'                                                      +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:507:in `each'                                                                        +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:507:in `invoke_before'                                                               +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:106:in `block in run_callbacks'                                                      +
                      | /var/www/canvas/app/controllers/application_controller.rb:484:in `report_to_datadog'                                                                                                     +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:117:in `block in run_callbacks'                                                      +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/inst_statsd-2.1.4/lib/inst_statsd/statsd.rb:95:in `batch'                                                                                  +
                      | /var/www/canvas/app/controllers/application_controller.rb:472:in `batch_statsd'                                                                                                          +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:117:in `block in run_callbacks'                                                      +
                      | /var/www/canvas/lib/temp_cache.rb:28:in `enable'                                                                                                                                         +
                      | /var/www/canvas/app/controllers/application_controller.rb:468:in `enable_request_cache'                                                                                                  +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:117:in `block in run_callbacks'                                                      +
                      | /var/www/canvas/app/controllers/application_controller.rb:462:in `set_locale'                                                                                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:117:in `block in run_callbacks'                                                      +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:135:in `run_callbacks'                                                               +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/abstract_controller/callbacks.rb:19:in `process_action'                                                             +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_controller/metal/rescue.rb:20:in `process_action'                                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'                                          +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/notifications.rb:166:in `block in instrument'                                                     +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/notifications/instrumenter.rb:21:in `instrument'                                                  +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/notifications.rb:166:in `instrument'                                                              +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_controller/metal/instrumentation.rb:30:in `process_action'                                                   +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_controller/metal/params_wrapper.rb:252:in `process_action'                                                   +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activerecord-5.1.6.2/lib/active_record/railties/controller_runtime.rb:22:in `process_action'                                               +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/abstract_controller/base.rb:124:in `process'                                                                        +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionview-5.1.6.2/lib/action_view/rendering.rb:30:in `process'                                                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_controller/metal.rb:189:in `dispatch'                                                                        +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_controller/metal.rb:253:in `dispatch'                                                                        +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/routing/route_set.rb:49:in `dispatch'                                                               +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/routing/route_set.rb:31:in `serve'                                                                  +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/journey/router.rb:50:in `block in serve'                                                            +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/journey/router.rb:33:in `each'                                                                      +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/journey/router.rb:33:in `serve'                                                                     +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/routing/route_set.rb:844:in `call'                                                                  +
                      | /var/www/canvas/gems/plugins/respondus_soap_endpoint/lib/respondus_soap_endpoint/middleware.rb:78:in `call'                                                                              +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/etag.rb:25:in `call'                                                                                                   +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/conditional_get.rb:25:in `call'                                                                                        +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/head.rb:12:in `call'                                                                                                   +
                      | /var/www/canvas/app/middleware/request_throttle.rb:60:in `block in call'                                                                                                                 +
                      | /var/www/canvas/app/middleware/request_throttle.rb:278:in `reserve_capacity'                                                                                                             +
                      | /var/www/canvas/app/middleware/request_throttle.rb:55:in `call'                                                                                                                          +
                      | /var/www/canvas/app/middleware/request_context_session.rb:25:in `call'                                                                                                                   +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/session/abstract/id.rb:232:in `context'                                                                                +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/session/abstract/id.rb:226:in `call'                                                                                   +
                      | /var/www/canvas/app/middleware/load_account.rb:29:in `call'                                                                                                                              +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/cookies.rb:613:in `call'                                                                 +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/callbacks.rb:26:in `block in call'                                                       +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/callbacks.rb:97:in `run_callbacks'                                                                +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/callbacks.rb:24:in `call'                                                                +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/debug_exceptions.rb:59:in `call'                                                         +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'                                                          +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/railties-5.1.6.2/lib/rails/rack/logger.rb:36:in `call_app'                                                                                 +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/railties-5.1.6.2/lib/rails/rack/logger.rb:26:in `call'                                                                                     +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'                                                                +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/request_store-1.4.1/lib/request_store/middleware.rb:19:in `call'                                                                           +
                      | /var/www/canvas/app/middleware/request_context_generator.rb:49:in `call'                                                                                                                 +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/method_override.rb:22:in `call'                                                                                        +
                      | /var/www/canvas/app/middleware/prevent_non_multipart_parse.rb:33:in `call'                                                                                                               +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/runtime.rb:22:in `call'                                                                                                +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/activesupport-5.1.6.2/lib/active_support/cache/strategy/local_cache_middleware.rb:27:in `call'                                             +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/actionpack-5.1.6.2/lib/action_dispatch/middleware/executor.rb:12:in `call'                                                                 +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/rack-2.0.6/lib/rack/sendfile.rb:111:in `call'                                                                                              +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/railties-5.1.6.2/lib/rails/engine.rb:522:in `call'                                                                                         +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/railties-5.1.6.2/lib/rails/railtie.rb:185:in `public_send'                                                                                 +
                      | /var/www/canvas/vendor/bundle/ruby/2.4.0/gems/railties-5.1.6.2/lib/rails/railtie.rb:185:in `method_missing'                                                                              +
                      | /usr/local/rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/passenger-6.0.2/src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb:97:in `process_request'                  +
                      | /usr/local/rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/passenger-6.0.2/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:149:in `accept_and_process_next_request'+
                      | /usr/local/rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/passenger-6.0.2/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:110:in `main_loop'                      +
                      | /usr/local/rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/passenger-6.0.2/src/ruby_supportlib/phusion_passenger/request_handler.rb:415:in `block (3 levels) in start_threads'             +
                      | /usr/local/rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/passenger-6.0.2/src/ruby_supportlib/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'
(1 row)
```

**Comment by qq7** [comment]
This is fixed now

**Comment by JedMeister** [comment]
This seems to be related to the RCE API (the `key must be 32 bytes` message appears to be related to a non-32 byte string used in the RCE API env), so should be resolved in v16.1 (if not v16.0).
